native/src/main/java/io/ballerina/stdlib/websocket/client/listener/SyncClientConnectorListener.java

```diff
@@ -168,7 +196,29 @@ public void onMessage(WebSocketBinaryMessage webSocketBinaryMessage) {
                                 ValueCreator.createTypedescValue(targetType));
                         break;
                 }
-                callback.complete(message);
+                if (message instanceof BError) {
+                    callback.complete(WebSocketUtil
+                            .createWebsocketError(String.format("data binding failed: %s", message),
+                                    WebSocketConstants.ErrorCode.Error));
+                } else if (this.targetType != null) {
+                    boolean validationEnabled = connectionInfo.getWebSocketEndpoint().getMapValue(
+                            WebSocketConstants.CLIENT_ENDPOINT_CONFIG).getBooleanValue(ENABLE_VALIDATION_CONFIG);
+                    if (validationEnabled) {
+                        Object validationResult = Constraints.validate(message, this.targetType);
+                        if (validationResult instanceof BError) {
+                            callback.complete(WebSocketUtil.createWebsocketErrorWithCause(
+                                    String.format("data validation failed: %s", validationResult),
+                                    WebSocketConstants.ErrorCode.PayloadValidationError,
+                                    (BError) validationResult));
+                        } else {
+                            callback.complete(message);
+                        }
+                    } else {
+                        callback.complete(message);
+                    }
```

Comment:
Can we move this to another function and reuse in L122 as well

Reply:
Fixed with a875f70